server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/InnerSpaceCompactionUtils.java

```diff
@@ -584,4 +587,46 @@ public static ICrossSpaceMergeFileSelector getCrossSpaceFileSelector(
       return new File[0];
     }
   }
+
+  /**
+   * Update the targetResource. Move xxx.target to xxx.tsfile and xxx.target.resource to
+   * xxx.tsfile.resource.
+   *
+   * @param targetResource the old tsfile to be moved, which is xxx.target
+   * @throws IOException
+   */
+  public static void moveTargetFile(TsFileResource targetResource, String fullStorageGroupName)
+      throws IOException {
+    if (!targetResource.getTsFilePath().endsWith(TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX)) {
+      logger.warn(
+          "TmpTargetFileReource "
+              + targetResource.getTsFilePath()
+              + " should be end with "
+              + TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX);
+      return;
+    }
+    File oldFile = targetResource.getTsFile();
+
+    // move TsFile and delete old tsfile
+    String newFilePath =
+        targetResource
+            .getTsFilePath()
+            .replace(TsFileNameGenerator.COMPACTION_TMP_FILE_SUFFIX, TsFileConstant.TSFILE_SUFFIX);
+    File newFile = new File(newFilePath);
+    FSFactoryProducer.getFSFactory().moveFile(oldFile, newFile);
+
+    // move resource file
+    targetResource.setFile(newFile);
+    targetResource.setClosed(true);
+    targetResource.serialize();
+
+    // delete old resource file
+    File oldResourceFile = new File(oldFile.getPath() + TsFileResource.RESOURCE_SUFFIX);
+    if (!oldResourceFile.delete()) {
+      logger.error(
+          "{} [Compaction] Fail to delete old resource file {}",
+          fullStorageGroupName,
+          oldResourceFile.getAbsolutePath());
```

Comment:
Use FSFactoryProducer.getFSFactory().moveFile(oldFile, newFile) to move resource, this is a atomic action.

Do not use write new, then delete. You will meet two .resource when restarting.